test/AutoDiff/Sema/DerivativeRegistrationCrossFile/main.swift

```diff
@@ -0,0 +1,9 @@
+// RUN: %target-swift-frontend -emit-sil -verify -primary-file %s %S/Inputs/derivatives.swift -module-name main -o /dev/null
+
+import _Differentiation
+
+@differentiable(reverse)
+func clamp(_ value: Double, _ lowerBound: Double, _ upperBound: Double) -> Double {
+    // No error expected
+    return max(min(value, upperBound), lowerBound)
```

Comment:
Could you please clarify the functionality added by this PR on top of https://github.com/apple/swift/pull/58644?

My understanding:
- https://github.com/apple/swift/pull/58644 adds an option for non-primary-file derivative resolution to `AbstractFunctionDecl::getDerivativeFunctionConfigurations`, as well as original declaration serialization for `@derivative` attribute.
- This PR moves the non-primary-file derivative resolution logic to `CompilerInstance::finishTypeChecking` instead, visiting each source file.
  - This resolves the "cross-file derivative not found" issue in https://github.com/apple/swift/issues/55170.
  - Does this also avoid circular requests?
  - Are there other implications to this new resolution-triggering condition? Ostensibly, to fix https://github.com/apple/swift/issues/55170, this PR is causing derivative resolution to happen in more cases than previously in https://github.com/apple/swift/pull/58644.

Reply:
1. Yes, as I mentioned, this avoids circular requests. As custom derivative resolution is decoupled from obtaining derivatives configurations. Also, after this PR the differentiation will not trigger typechecking
2. After this PR indeed, we lookup and register all explicit derivates regardless of their original function and regardless whether this function is used. The nice side effect is that we will error out in case of `@derivative(of: foo)` with `foo` not defined anywhere.
3. Effectively this resolves the following cross-file derivate requests:
   *  File A.swift is using (wants a derivative of) function from B.swift with derivatives defined in C.swift
   * The important special case is when B.swift is effectively a part of stdlib (like in the testcase). Here C.swift is a non-primary source but we have no way to resolve the derivative in other way as there is no mechanism to "pull" anything from C.swift – there are simply no forward dependency edges in the dependency graph, only backward.